ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java

```diff
@@ -188,16 +199,61 @@ && getUserRateLimiterMap().get(user.getName()) != null
             if (getMlGuard() != null && !getMlGuard().validate(payload, MLGuard.Type.INPUT)) {
                 throw new IllegalArgumentException("guardrails triggered for user input");
             }
-            invokeRemoteModel(mlInput, parameters, payload, tensorOutputs, countDownLatch, actionListener);
+            if (executionContext.getConnectorRetryOption().isRetryEnabled()) {
+                invokeRemoteModelWithRetry(mlInput, parameters, payload, executionContext, actionListener);
+            } else {
+                invokeRemoteModel(mlInput, parameters, payload, executionContext, actionListener);
+            }
         }
     }
 
+    default void invokeRemoteModelWithRetry(
+        MLInput mlInput,
+        Map<String, String> parameters,
+        String payload,
+        ExecutionContext executionContext,
+        ActionListener<Tuple<Integer, ModelTensors>> actionListener
+    ) {
+        final RetryableAction<Tuple<Integer, ModelTensors>> invokeRemoteModelAction = new RetryableAction<>(
+            getLogger(),
+            getClient().threadPool(),
+            TimeValue.timeValueMillis(executionContext.getConnectorRetryOption().getRetryBackoffMillis()),
+            TimeValue.timeValueSeconds(executionContext.getConnectorRetryOption().getRetryTimeoutSeconds()),
+            actionListener,
+            BackoffPolicy
+                .constantBackoff(
+                    TimeValue.timeValueMillis(executionContext.getConnectorRetryOption().getRetryBackoffMillis()),
+                    Integer.MAX_VALUE
```

Comment:
I think we should have a max number of retries to not always time_out in edge cases. Imagine for a fully throttled service and we need to time out every request causing strange performance behaviors.

Reply:
For the cases you mentioned this settings make sense. But it is hard to set a value proper for all cases. The value is infected by input doc length, remote endpoint type, retry backoff value, and even the workload amount from other services if they share the same SageMaker endpoint with OpenSearch.
I think we can implement if as @xinyual suggest: https://github.com/opensearch-project/ml-commons/pull/2462#discussion_r1609139231
We add this setting, but disable this setting by default